core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.iceberg.ManifestGroup.CreateTasksFunction;
+import org.apache.iceberg.ManifestGroup.TaskContext;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.FluentIterable;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.SnapshotUtil;
+import org.apache.iceberg.util.TableScanUtil;
+
+class BaseIncrementalChangelogScan
+    extends BaseIncrementalScan<
+        IncrementalChangelogScan, ChangelogScanTask, ScanTaskGroup<ChangelogScanTask>>
+    implements IncrementalChangelogScan {
+
+  BaseIncrementalChangelogScan(TableOperations ops, Table table) {
+    this(ops, table, table.schema(), new TableScanContext());
+  }
+
+  BaseIncrementalChangelogScan(
+      TableOperations ops, Table table, Schema schema, TableScanContext context) {
+    super(ops, table, schema, context);
+  }
+
+  @Override
+  protected IncrementalChangelogScan newRefinedScan(
+      TableOperations newOps, Table newTable, Schema newSchema, TableScanContext newContext) {
+    return new BaseIncrementalChangelogScan(newOps, newTable, newSchema, newContext);
+  }
+
+  @Override
+  protected CloseableIterable<ChangelogScanTask> doPlanFiles(
+      Long fromSnapshotIdExclusive, long toSnapshotIdInclusive) {
+
+    Deque<Snapshot> changelogSnapshots =
```

Comment:
Opt/Nit:  It seems a bit strange to have a deque so early, when I see its really needed inside CreateDataFileChangeTasks.  You have to kind of track the code in two places to see how it works.

What do you think about, just using Iterable<Snapshot>  here and then only doing the Deque / computeSnapshotOrdinals logic inside the  CreateDataFileChangeTasks constructor?

The slight negative might be , having to transform Iterable<Snapshot> to stream for some of these methods.

Reply:
I used `Deque` cause I need a collection of ordered snapshots. I thought about using `List` with `LinkedList` as the underlying implementation but `LinkedList` is less efficient than `ArrayDeque`.

I could use `Deque` in the method but return `Collection` but will that be cleaner?

Reply:
Yea , or maybe just clarify the variable or method, that this is in order ascending (by timestamp), and that's what the CreateDataFileChangeTasks expects.

Or how about just create the ordinals in this method, and have CreateDataFileChangeTasks take in the ordinals?

I guess the whole goal for me would be to make it easier to see that that the ordinals are in order ascending without having to jump around different methods/ inner classes

Reply:
Got it. Let me see what I can do.

Reply:
I renamed the method and added a comment on what the order is and why it is important.